api/include/opentelemetry/metrics/instrument.h

```diff
@@ -0,0 +1,191 @@
+enum class InstrumentKind
+{
+  IntCounter,
+  IntUpDownCounter,
+  IntValueRecorder,
+  IntValueObserver,
+  IntSumObserver,
+  IntUpDownSumObserver,
+  DoubleCounter,
+  DoubleUpDownCounter,
+  DoubleValueRecorder,
+  DoubleValueObserver,
+  DoubleSumObserver,
+  DoubleUpDownSumObserver
+};
+
+// Fewer Bound types because Asynchronous instruments cannot bind
+enum class BoundInstrumentKind
+{
+  BoundIntCounter,
+  BoundIntUpDownCounter,
+  BoundIntValueRecorder,
+  BoundDoubleCounter,
+  BoundDoubleUpDownCounter,
+  BoundDoubleValueRecorder
+};
+
+class Instrument
+{
+
+public:
+  // Note that Instruments should be created using the Meter class.
+  // Please refer to meter.h for documentation.
+  Instrument() = default;
+
+  /**
+   * Base class constructor for all other instrument types.  Whether or not
+   * an instrument is synchronous or bound, it requires a name, description,
+   * unit, and enabled flag.
+   *
+   * @param name is the identifier of the instrumenting library
+   * @param description explains what the metric captures
+   * @param unit specifies the data type held in the instrument
+   * @param enabled determines if the metric is currently capturing data
+   * @return Instrument type with the specified attirbutes
+   */
+  Instrument(nostd::string_view name,
+             nostd::string_view description,
+             nostd::string_view unit,
+             bool enabled)
+  {}
+
+  // Returns true if the instrument is enabled and collecting data
+  virtual bool IsEnabled()
+  {
+    // False in default implementation
+    return false;
+  }
+
+  // Return the instrument name
+  virtual nostd::string_view GetName() { return nostd::string_view(""); }
+
+  // Return the instrument description
+  virtual nostd::string_view GetDescription() { return nostd::string_view(""); }
+
+  // Return the insrument's units of measurement
+  virtual nostd::string_view GetUnits() { return nostd::string_view(""); }
+
+  virtual ~Instrument() = default;
+};
+
+class BoundSynchronousInstrument : public Instrument
+{
+
+public:
+  BoundSynchronousInstrument() = default;
+
+  BoundSynchronousInstrument(nostd::string_view name,
+                             nostd::string_view description,
+                             nostd::string_view unit,
+                             bool enabled)
+  {}
+
+  /**
+   * Frees the resources associated with this Bound Instrument.
+   * The Metric from which this instrument was created is not impacted.
+   *
+   * @param none
+   * @return void
+   */
+  virtual void unbind() {}
+
+  /**
+   * Records a single synchronous metric event; a call to the aggregator
+   * Since this is a bound synchronous instrument, labels are not required in  * metric capture
+   * calls.
+   *
+   * @param value is the numerical representation of the metric being captured
+   * @return void
+   */
+  virtual void update(nostd::variant<int, double> value) final {}
```

Comment:
I don't see a situation where you would want to change the type you are calling update with after constructing the instrument, would it make sense to template the instrument class and use its type here? That way you don't have any overhead at runtime to unwrap the value.

Reply:
Our initial hesitance with templates was that they aren't restrictive enough given that we only want to support int and double types.  That said, I think you make a great point and I'll reevaluate with my team as it does reduce overhead and cut down on duplicate code.  

If I were to template, would I then be responsible for checking that the user picks a suitable type (perhaps by using something like <type_traits>) or expect them to use the instruments with only ints and doubles as intended?

Reply:
Using traits could be one way to go. The compiler errors for using an unsupported type would be pretty gross in this situation, I don't know how much we want to worry about that though.

Alternatively, you could break them out into their own classes with specific types they accept, like `DoubleInstrument`, and `IntInstrument`. Code duplication in that case seems pretty rough.

We definitely want to help the user out where possible by putting some guard rails up on the types they can use, one way or another.